### ISSUE TITLE
aws_cloudwatch_log_metric_filter creation needs to be serial

### ISSUE THREAD

**Issue by lxop** [terraform-providers/terraform-provider-aws#7605]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.11.10
+ provider.aws v1.42.0
+ provider.cloudflare v1.7.0
+ provider.template v1.0.0
```

### Affected Resource(s)

* aws_cloudwatch_log_metric_filter

### Terraform Configuration Files

```hcl
resource "aws_cloudwatch_log_metric_filter" "ssh_login_user" {
  count = "${length(var.ssh_users)}"
  name = "Session-opened-for-user-${var.ssh_users[count.index]}"
  pattern        = "session opened for user ${var.ssh_users[count.index]}"
  log_group_name = "${var.log_group_name}"

  metric_transformation {
    name      = "AWSAgentSSHLoginEventCount-${var.ssh_users[count.index]}"
    namespace = "AWSLoggingAgentLogs"
    value     = "1"
  }
}
```

### Expected Behavior

A bunch of metric filters are created

### Actual Behavior

Some metric filters are created, some are not, and error messages are issued:
```
3 error(s) occurred:

* (path).ssh_login_user[8]: 1 error(s) occurred:

* aws_cloudwatch_log_metric_filter.ssh_login_user.8: Creating/Updating CloudWatch Log Metric Filter failed: OperationAbortedException: A conflicting operation is currently in progress against this resource. Please try again.
	status code: 400, request id: 59481c0e-3465-11e9-b058-35cd0d0946e2
* (path).ssh_login_user[6]: 1 error(s) occurred:

* aws_cloudwatch_log_metric_filter.ssh_login_user.6: Creating/Updating CloudWatch Log Metric Filter failed: OperationAbortedException: A conflicting operation is currently in progress against this resource. Please try again.
	status code: 400, request id: 5947ce0c-3465-11e9-8052-93b7d847ef42
* (path).ssh_login_user[5]: 1 error(s) occurred:

* aws_cloudwatch_log_metric_filter.ssh_login_user.5: Creating/Updating CloudWatch Log Metric Filter failed: OperationAbortedException: A conflicting operation is currently in progress against this resource. Please try again.
	status code: 400, request id: 5949ca85-3465-11e9-a0ba-89bfabc12fd6
```
Rerunning `terraform apply` produces the expected result.

### Steps to Reproduce

1. Have several metric filters to provision. In the case of the hcl above, this would mean having several names in the `ssh_users` variable. "Several" in our case is 9.
2. `terraform apply`
3. Sometimes (often, but not always), some filters will fail and will produce an error message

### References

* https://github.com/hashicorp/terraform/issues/10802
* #789

**Comment by lxop** [comment]
I'm having a stab at this

**Comment by batzel** [comment]
Ran into this today. I don't think it's that the operations need to be done serially -- looks like some sort of internal ordering issue in terraform? 
1st workaround: "terraform apply -parallelism=1" seems to "work" every time. But...
2nd problem: Every time I run 'terraform plan' I'm seeing attempts to change resources:

`  ~ module.foo.aws_cloudwatch_log_metric_filter.cis3-1
      metric_transformation.0.name: "KMSCustomerKeyDeletion" => "UnauthorizedAttemptCount"
      pattern:                      "{ ($.eventSource = kms.amazonaws.com) && (($.eventName=DisableKey) || ($.eventName=ScheduleKeyDeletion)) }" => "{ ($.errorCode = \"*UnauthorizedOperation\") || ($.errorCode = \"AccessDenied*\") }"

  ~ module.foo.aws_cloudwatch_log_metric_filter.cis3-6
      metric_transformation.0.name: "KMSCustomerKeyDeletion" => "ConsoleLoginFailures"
      pattern:                      "{ ($.eventSource = kms.amazonaws.com) && (($.eventName=DisableKey) || ($.eventName=ScheduleKeyDeletion)) }" => "{ ($.eventName = ConsoleLogin) && ($.errorMessage = \"Failed authentication\") }"


Plan: 0 to add, 8 to change, 0 to destroy.
`
terraform apply -parallelism=1 out.plan outputs:
`module.foo.aws_cloudwatch_log_metric_filter.cis3-1: Modifying... (ID: UnauthorizedApiCallsFilter)
  metric_transformation.0.name: "KMSCustomerKeyDeletion" => "UnauthorizedAttemptCount"
  pattern:                      "{ ($.eventSource = kms.amazonaws.com) && (($.eventName=DisableKey) || ($.eventName=ScheduleKeyDeletion)) }" => "{ ($.errorCode = \"*UnauthorizedOperation\") || ($.errorCode = \"AccessDenied*\") }"
module.foo.aws_cloudwatch_log_metric_filter.cis3-1: Modifications complete after 0s (ID: UnauthorizedApiCallsFilter)
module.foo.aws_cloudwatch_log_metric_filter.cis3-6: Modifying... (ID: UnauthorizedApiCallsFilter)
  metric_transformation.0.name: "KMSCustomerKeyDeletion" => "ConsoleLoginFailures"
  pattern:                      "{ ($.eventSource = kms.amazonaws.com) && (($.eventName=DisableKey) || ($.eventName=ScheduleKeyDeletion)) }" => "{ ($.eventName = ConsoleLogin) && ($.errorMessage = \"Failed authentication\") }"
module.foo.aws_cloudwatch_log_metric_filter.cis3-6: Modifications complete after 0s (ID: UnauthorizedApiCallsFilter)
...
Apply complete! Resources: 0 added, 8 changed, 0 destroyed.
`
And another terraform plan -parallelism=1 -out out.plan wants to move things around again:
`  ~ module.foo.aws_cloudwatch_log_metric_filter.cis3-1
      metric_transformation.0.name: "ConsoleLoginFailures" => "UnauthorizedAttemptCount"
      pattern:                      "{ ($.eventName = ConsoleLogin) && ($.errorMessage = \"Failed authentication\") }" => "{ ($.errorCode = \"*UnauthorizedOperation\") || ($.errorCode = \"AccessDenied*\") }"

  ~ module.foo.aws_cloudwatch_log_metric_filter.cis3-7
      metric_transformation.0.name: "ConsoleLoginFailures" => "KMSCustomerKeyDeletion"
      pattern:                      "{ ($.eventName = ConsoleLogin) && ($.errorMessage = \"Failed authentication\") }" => "{ ($.eventSource = kms.amazonaws.com) && (($.eventName=DisableKey) || ($.eventName=ScheduleKeyDeletion)) }"

...
Plan: 0 to add, 8 to change, 0 to destroy.
`
Note that it was my cis3-1 and cis3-6 rules first time, then cis3-1 and cis3-7 the next time.

**Comment by lxop** [comment]
@batzel try my branch and see if that fixes your issue

**Comment by bflad** [comment]
The fix for this has been merged and will release with version 2.2.0 of the Terraform AWS Provider, likely later today.

**Comment by bflad** [comment]
This has been released in [version 2.2.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#220-march-15-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
